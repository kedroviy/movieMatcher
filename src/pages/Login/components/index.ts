export { LoginAuth } from './login-auth';
export { LoginRegistration } from './login-registration';
