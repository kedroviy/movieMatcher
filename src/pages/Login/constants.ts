export const STRINGS = {
    WELCOME: 'Добро пожаловать в MovieMatch',
    SUB_WELCOME: 'Сделай подбор фильмов самостоятельно или в компании друзей!',
    PASSWORD_RECOVERY: 'Восстановить пароль',
}