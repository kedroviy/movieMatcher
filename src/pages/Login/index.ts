export * from './components';
export { LoginScreen } from './Login';
