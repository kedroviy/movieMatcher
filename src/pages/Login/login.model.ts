export type RecoveryPasswordType = {
    password: string,
    confirmPassword: string,
    isFormValidPassword: boolean,
    isFormValidConfirmPassword: boolean,
};

export type RecoveryPasswordActionType =
    | { type: 'SET_PASSWORD'; payload: string }
    | { type: 'SET_CONFIRM_PASSWORD'; payload: string };
