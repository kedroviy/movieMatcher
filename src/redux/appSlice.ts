import { createSlice } from "@reduxjs/toolkit";

type AppState = {
    appVersion: string;
    error: null | string;
    loading: boolean;
    loadingApplication: boolean;
    success: boolean;
    onboarded: boolean;
}

const initialState: AppState = {
    appVersion: '1.0',
    error: null,
    loading: false,
    loadingApplication: false,
    success: false,
    onboarded: false,
};

const appSlice = createSlice({
    name: 'app',
    initialState,
    reducers: {
        
    },
});

export default appSlice.reducer;