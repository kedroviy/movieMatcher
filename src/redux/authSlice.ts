import { createAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as Keychain from 'react-native-keychain';

import { removeToken, saveToken } from '../shared';
import {
    loginUser,
    registrationUser,
    sendGoogleCodeToServer,
} from 'features';
import { GoogleSignin } from '@react-native-google-signin/google-signin';

type AuthState = {
    token: string | null;
    isAuthenticated: boolean;
    googleAuthCode: string | null;
    error: null | string;
    loading: boolean;
    loadingApplication: boolean;
    success: boolean;
}

const initialState: AuthState = {
    token: null,
    isAuthenticated: false,
    googleAuthCode: null,
    error: null,
    loading: false,
    loadingApplication: false,
    success: false,
};

GoogleSignin.configure({
    webClientId: '277043180680-ugnrupqacuchkhpklb1qhv8kaueh4eks.apps.googleusercontent.com',
    offlineAccess: true,
    forceCodeForRefreshToken: true,
});

export const authUser = createAsyncThunk(
    'auth/LOGIN',
    async (userData: { email: string; password: string }, { rejectWithValue }) => {
        try {
            const response = await loginUser(userData);

            if (response.success) {
                saveToken(response.token as string);
            }
            return response;
        } catch (error) {
            return rejectWithValue(error);
        }
    }
);

export const authenticateWithGoogle = createAsyncThunk(
    'auth/GOOGLE',
    async (_, { rejectWithValue }) => {
        try {
            await GoogleSignin.hasPlayServices();
            const userInfo = await GoogleSignin.signIn();
            const idToken = userInfo.idToken;

            const response = await sendGoogleCodeToServer(idToken as string);

            if (response.success) {
                saveToken(response.token as string);
            }

            return response;
        } catch (error) {
            return rejectWithValue(error || 'An error occurred');
        }
    });

export const authRegistrationUser = createAsyncThunk(
    'auth/REGISTRATION',
    async (userData: { email: string; password: string }, { rejectWithValue }) => {
        try {
            const response = await registrationUser(userData);
            if (!response.success) {
                return rejectWithValue(response?.error);
            }
            return response
        } catch (error) {
            return rejectWithValue(error);
        }
    }
);

export const checkAuthStatus = createAsyncThunk(
    'auth/CHECK_STATUS',
    async () => {
        try {
            const credentials = await Keychain.getGenericPassword({ service: 'token_guard' });
            if (credentials) {
                return { isAuthenticated: true };
            }
            return { isAuthenticated: false };
        } catch (error) {
            return { isAuthenticated: false };
        }
    }
);

// export const sendRecoveryCodeEffect = createAsyncThunk(
//     'auth/SEND_RECOVERY_CODE',
//     async (email: string, { rejectWithValue }) => {
//         try {
//             const response = await sendRecoveryCodeAPI(email);

//             if (response.success) {
//                 console.log(response.success);
//             }

//             return response;
//         } catch (error) {
//             return rejectWithValue(error);
//         }
//     }
// );

export const saveGoogleAuthCode = createAction<string>('auth/SAVE_GOOGLE_AUTH_CODE');

export const revertAll = createAction('REVERT_ALL');

const authSlice = createSlice({
    name: 'auth',
    initialState,
    reducers: {
        reset: () => initialState,
        logout: (state) => {
            state.isAuthenticated = false;
            state.token = null;
            removeToken();
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(checkAuthStatus.pending, (state) => {
                state.loading = true;
                state.loadingApplication = true;
            })
            .addCase(checkAuthStatus.fulfilled, (state, action) => {
                state.loading = false;
                state.isAuthenticated = action.payload.isAuthenticated;
                state.loadingApplication = false;
            })
            .addCase(checkAuthStatus.rejected, (state) => {
                state.loading = false;
                state.error = "Ошибка при проверке статуса аутентификации";
                state.loadingApplication = false;
            })
            .addCase(authUser.pending, (state) => {
                state.loading = true;
            })
            .addCase(authUser.fulfilled, (state, action) => {
                state.loading = false;
                state.token = action.payload.token as string;

                if (action.payload.success) {
                    state.token = action.payload.token as string;
                    state.isAuthenticated = true;
                } else {
                    state.isAuthenticated = false;
                    state.token = null;
                }
            })
            .addCase(authUser.rejected, (state, action) => {
                state.loading = false;
                state.error = action.payload as string;
            })
            .addCase(authenticateWithGoogle.pending, (state) => {
                state.loading = true;
            })
            .addCase(authenticateWithGoogle.fulfilled, (state, action) => {
                state.loading = false;

                if (action.payload.success) {
                    state.isAuthenticated = true;
                } else {
                    state.isAuthenticated = false;
                    state.token = null;
                }
            })
            .addCase(authenticateWithGoogle.rejected, (state, action) => {
                state.loading = false;
                state.error = action.payload as string;
            })
            .addCase(authRegistrationUser.pending, (state) => {
                state.loading = true;
            })
            .addCase(authRegistrationUser.fulfilled, (state, action) => {
                state.loading = false;
            })
            .addCase(authRegistrationUser.rejected, (state, action) => {
                state.loading = false;
                state.error = action.payload as string;
            })
            .addCase(saveGoogleAuthCode, (state, action) => {
                state.googleAuthCode = action.payload;
            })
    },
});

export const { logout } = authSlice.actions;
export default authSlice.reducer;
