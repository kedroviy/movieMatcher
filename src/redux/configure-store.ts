import { configureStore } from '@reduxjs/toolkit';
import authSliceReducer from './authSlice';
import recoveryPasswordReducer from './recoveryPasswordSlice';
import appSliceReducer from './appSlice';
import userSliceReducer from './userSlice';
import moviesSliceReducer from './moviesSlice';
import matchSliceReducer from './matchSlice';

export const store = configureStore({
  reducer: {
    authSlice: authSliceReducer,
    recoveryPasswordSlice: recoveryPasswordReducer,
    appSlice: appSliceReducer,
    userSlice: userSliceReducer,
    moviesSlice: moviesSliceReducer,
    matchSlice: matchSliceReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
