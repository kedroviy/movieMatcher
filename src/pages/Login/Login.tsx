import { useDispatch, useSelector } from "react-redux";
import { StyleSheet, Dimensions, Text, TouchableOpacity } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { View } from 'react-native-ui-lib';

import StartLogotype from '../../../assets/startLogo.svg';
import GoogleIcon from '../../../assets/google.svg'

import { AppDispatch } from "../../redux/configure-store";
import { authenticateWithGoogle } from "../../redux/authSlice";
import { Loader } from "../../shared";

export const LoginScreen = () => {
    const dispatch: AppDispatch = useDispatch();
    const { loading } = useSelector((state: any) => state.authSlice);
    const navigation: any = useNavigation();
    const windowWidth = Dimensions.get('window').width;

    const onAuthWithGoogle = async () => {
        await dispatch(authenticateWithGoogle());
    };

    return (
        <View style={[styles.container, { width: windowWidth }]}>
            <View style={{ width: 328, height: 328, marginVertical: 25, }}>
                <StartLogotype />
            </View>
            <View style={{ width: 328, gap: 12, alignItems: 'center', marginTop: 20 }}>
                <Text style={{
                    color: '#FFFFFF',
                    fontSize: 22,
                    fontWeight: '700',
                    lineHeight: 27.5,
                    textAlign: 'center'
                }}>Добро пожаловать в MovieMatch!</Text>
                <Text style={{
                    color: '#FFFFFF',
                    fontSize: 16,
                    fontWeight: '400',
                    lineHeight: 21.6,
                    textAlign: 'center'
                }}>Сделай подбор фильмов самостоятельно или в компании друзей</Text>
            </View>
            <View style={{ width: '100%', alignItems: 'center', gap: 16, marginTop: 52, }}>
                <TouchableOpacity
                    style={[
                        styles.button,
                        {
                            flexDirection: 'row',
                            width: windowWidth - 32,
                            height: 48,
                            backgroundColor: '#DF0A1E',
                            alignItems: 'center',
                            justifyContent: 'center',
                        }
                    ]}
                    onPress={() => navigation.navigate('LoginAuth')}
                >
                    <Text style={{
                        color: '#FFF',
                        fontSize: 18,
                        fontWeight: '500',
                    }}>Войти</Text>
                </TouchableOpacity>
                <TouchableOpacity
                    style={[
                        styles.button,
                        {
                            flexDirection: 'row',
                            width: windowWidth - 32,
                            height: 48,
                            backgroundColor: '#FFF',
                            alignItems: 'center',
                            justifyContent: 'flex-start',
                        }
                    ]}
                    onPress={onAuthWithGoogle}
                >
                    <GoogleIcon width={36} height={36} />
                    <Text style={{
                        color: '#1D1D1D',
                        fontSize: 18,
                        fontWeight: '500',
                        left: (windowWidth / 6),
                    }}>Войти с Google</Text>
                </TouchableOpacity>
                <TouchableOpacity
                    style={[
                        styles.button,
                        {
                            flexDirection: 'row',
                            width: windowWidth - 32,
                            height: 48,
                            backgroundColor: '#4AB34C',
                            alignItems: 'center',
                            justifyContent: 'center',
                        }
                    ]}
                    onPress={() => navigation.navigate('LoginRegistration')}
                >
                    <Text style={{
                        color: '#FFF',
                        fontSize: 18,
                        fontWeight: '500',
                    }}>Зарегистрироваться</Text>
                </TouchableOpacity>
            </View>
            {loading ? <Loader /> : null}
        </View>
    )
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: '#353535',
        flex: 1,
        alignItems: 'center',
    },
    button: {
        alignItems: 'center',
        backgroundColor: '#595959',
        padding: 10,
        width: 328,
        height: 44,
        gap: 10,
        borderRadius: 5,
    },
});
