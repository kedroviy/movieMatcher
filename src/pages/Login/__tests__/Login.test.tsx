import 'react-native';
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { LoginScreen } from '../Login';
import { useDispatch, useSelector } from 'react-redux';
import { Dispatch } from 'redux';
import { it } from '@jest/globals';
import renderer from 'react-test-renderer';
import { TypedUseSelectorHook } from 'react-redux';

type MockDispatchFn = typeof useDispatch & jest.Mock;
type MockSelectorFn<TState = unknown> = TypedUseSelectorHook<TState> & jest.Mock;

const mockNavigation = {
    navigate: jest.fn(),
};

jest.mock('@react-navigation/native', () => ({
    useNavigation: () => mockNavigation,
}));

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(() => jest.fn()),
    useSelector: jest.fn(),
}));

const mockDispatch: MockDispatchFn = useDispatch as any;
const mockSelector: MockSelectorFn = useSelector as any;

beforeEach(() => {
    jest.clearAllMocks();

    (mockDispatch as jest.Mock).mockReturnValue({
        loading: false,
        error: null,
        isAuthenticated: false,
        loadingApplication: false,
    });

    (mockSelector as jest.Mock).mockReturnValue(jest.fn());
});


describe('LoginScreen Component Testing', () => {
    it('LoginScreen renders correctly', () => {
        const tree = renderer.create(<LoginScreen />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});