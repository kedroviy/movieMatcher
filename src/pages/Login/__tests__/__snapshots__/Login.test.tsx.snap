// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen Component Testing LoginScreen renders correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#353535",
        "flex": 1,
      },
      {
        "width": 750,
      },
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "height": 328,
        "marginVertical": 24,
        "width": 750,
      }
    }
  >
    <mocked-file />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "gap": 12,
        "marginTop": 32,
        "width": 328,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 22,
          "fontWeight": "700",
          "lineHeight": 27.5,
          "textAlign": "center",
        }
      }
    >
      Добро пожаловать в MovieMatch
    </Text>
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 21.6,
          "textAlign": "center",
        }
      }
    >
      Сделай подбор фильмов самостоятельно или в компании друзей!
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "gap": 16,
        "marginTop": 32,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ED0E0E",
          "borderRadius": 5,
          "flexDirection": "row",
          "gap": 10,
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 10,
          "width": 718,
        }
      }
      testID="myButton"
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 18,
            "fontWeight": "500",
          }
        }
      >
        Войти
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFF",
          "borderRadius": 5,
          "flexDirection": "row",
          "gap": 10,
          "height": 48,
          "justifyContent": "flex-start",
          "opacity": 1,
          "padding": 10,
          "width": 718,
        }
      }
      testID="myButton"
    >
      <mocked-file
        height={36}
        width={36}
      />
      <Text
        style={
          {
            "color": "#1D1D1D",
            "fontSize": 18,
            "fontWeight": "500",
            "left": 125,
          }
        }
      >
        Войти с Google
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#4AB34C",
          "borderRadius": 5,
          "flexDirection": "row",
          "gap": 10,
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 10,
          "width": 718,
        }
      }
      testID="myButton"
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 18,
            "fontWeight": "500",
          }
        }
      >
        Зарегистрироваться
      </Text>
    </View>
  </View>
</View>
`;
